### PR TITLE
Improve email address validation behaviour

### PR DESCRIPTION
**Pull request by stevebaxter**
This adds a custom validation message when trying to create an account if it already exists. We tell the user how the original account was created.

This should also fix a problem some users may have when trying to log in with FB or Twitter if there are duplicate email addresses in the system.

fixes: #353 

Looks a bit like this:

Screenshot 2019-12-11 15 56 50

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -27,7 +27,9 @@ class User < ApplicationRecord
   after_save :send_welcome_email, if: :needs_welcome_email?
   before_destroy :clear_swap
 
-  validates :email, uniqueness: { case_sensitive: false }, allow_nil: true
+  # Additional validation for emails, :validatable has already added basic validation
+  validate :email_uniqueness, on: :create
```
Comment (stevebaxter) [inline app/models/user.rb]: We only validate on create, which should prevent some of the problems we're seeing. `:validatable` only validates on change:

```
        base.class_eval do
          validates_presence_of   :email, if: :email_required?
          if Devise.activerecord51?
            validates_uniqueness_of :email, allow_blank: true, case_sensitive: true, if: :will_save_change_to_email?
            validates_format_of     :email, with: email_regexp, allow_blank: true, if: :will_save_change_to_email?
          else
            validates_uniqueness_of :email, allow_blank: true, if: :email_changed?
            validates_format_of     :email, with: email_regexp, allow_blank: true, if: :email_changed?
          end

          validates_presence_of     :password, if: :password_required?
          validates_confirmation_of :password, if: :password_required?
          validates_length_of       :password, within: password_length, allow_blank: true
        end
      end
```

File: app/models/user.rb [hunk 2]
```diff
@@ -321,4 +323,29 @@ def test_user_suffix
   def password_required?
     false
   end
+
+  def email_uniqueness
```
Comment (stevebaxter) [inline app/models/user.rb]: Custom validator. There may be a terser way to do this...

File: app/models/user.rb [hunk 3]
```diff
@@ -321,4 +323,29 @@ def test_user_suffix
...
+    end
+
+    return unless existing_user
+
+    # Delete the default message that gets added
+    errors.delete(:email)
+
+    # Add our custom one
+    if existing_user.email_login?
+      errors.add(:base, "A user with this email address already exists")
```
Comment (stevebaxter) [inline app/models/user.rb]: @tdg - are these messages ok?

File: app/models/user.rb [hunk 4]
```diff
@@ -321,4 +323,29 @@ def test_user_suffix
...
     false
   end
+
+  def email_uniqueness
+    # Ignore nil email addresses
+    return unless email
+
+    if id
+      # Ignore self in the uniqueness check
+      existing_user = User.where("lower(#{:email}) = ? and id <> ?", email.downcase, id).first
```
Comment (aspiers) [inline app/models/user.rb]:
```suggestion
      existing_user = User.find_by("lower(#{:email}) = ? and id <> ?", email.downcase, id)
```

File: app/models/user.rb (hunk 3 above)
Comment (aspiers) [reply app/models/user.rb]: Fine by me

**Comment by stevebaxter** [comment]
Merged from master and resolved conflicts

**Pull request merged**